naga/src/compact/functions.rs

```diff
@@ -30,10 +32,24 @@ impl FunctionTracer<'_> {
             }
         }
 
+        let mut typifier = crate::front::Typifier::new();
+
         // Treat named expressions as alive, for the sake of our test suite,
-        // which uses `let blah = expr;` to exercise lots of things.
+        // which uses `let blah = expr;` to exercise lots of things. The
+        // exception is expressions of abstract types, which can occur for local
+        // constants, which we must remove so that abstract types do not reach
+        // the IR.
         for (&value, _name) in &self.function.named_expressions {
-            self.expressions_used.insert(value);
+            typifier
+                .grow(value, &self.function.expressions, &self.resolve_context)
+                .expect("Failed to resolve type for named expression");
```

Comment:
perf: @jamienicol was concerned about this being slow, but @teoxoy doesn't think we should be worried about it.